core/trino-main/src/main/java/io/trino/sql/planner/plan/AggregationNode.java

```diff
@@ -54,6 +54,10 @@ public class AggregationNode
     private final Optional<Symbol> hashSymbol;
     private final Optional<Symbol> groupIdSymbol;
     private final List<Symbol> outputs;
+    /**
+     * Indicates whether aggregation is potentially reducing rows that are propagated though exchange operator.
+     */
+    private final Optional<Boolean> exchangeInputAggregation;
```

Comment:
I see. That's brittle. There are no guarantees that the node will remain to be the input of an exchange if other rules push things below the join afterwards, for instance.

If we can model it solely based on properties of the aggregation (e.g., whether it reduces the size of its input, as you suggested) that would be better.

Reply:
> We don't want to always keep intermediate aggregations above join as it would lead to multiple intermediate aggregations if there are many joins on top of each other. This is causing significant regressions.

Ideally, shouldn't adaptive partial aggregation avoid big regressions in this case ? Is this happening because partial/intermediate aggregation has significant cost even in the "disabled" state ?

Reply:
How about we rename this to `isInputReducingAggregation` ?

Reply:
As I understood it, the optimizer might be wrong about the decision to push down partial aggregation through join, so we want to retain a partial aggregation on top of the join while pushing a partial aggregation through. The simple approach would be to leave an intermediate aggregation on top of every join after pushdown. But when we have multiple joins on top of each other in a fragment, this leads to many intermediate aggregations. Ideally, these should adaptively turn off at runtime, but this is still not free and leads to significant regressions. Hence the extra field in `AggregationNode` to retain intermediate aggregation only on the join before remote data exchange. There's no regressions when benchmarking with this approach.
I would define `isInputReducingAggregation` as an auxiliary aggregation step introduced as a hedge against potentially non-optimal decision to push down partial aggregation more aggressively.